systemtest/infra_kafka.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return err
+	}
+	namespaceVar := tfexec.Var(fmt.Sprintf("namespace=%s", cfg.Namespace))
+	nameVar := tfexec.Var(fmt.Sprintf("name=%s", cfg.Name))
+	topicsVar := tfexec.Var(fmt.Sprintf("topics=%s", jsonTopics))
+	// Ensure terraform destroy runs once per TF path.
+	RegisterDestroy(cfg.TFPath, func() {
+		logger().Info("destroying provisioned Kafka infrastructure...")
+		tf.Destroy(ctx, topicsVar, namespaceVar, nameVar)
+	})
+	if err := tf.Apply(ctx, topicsVar, namespaceVar, nameVar); err != nil {
+		return err
+	}
+	tfOutput, err := tf.Output(ctx)
+	if err != nil {
+		return err
+	}
+	if raw, ok := tfOutput[KafkaBrokersKey]; ok {
+		var brokers []string
+		if err := json.Unmarshal(raw.Value, &brokers); err != nil {
+			return err
+		}
+		if len(brokers) > 0 {
+			SetKafkaBrokers(brokers...)
+		}
+	}
+	dt := string(bytes.TrimSpace(tfOutput[KafkaDeploymentTypeKey].Value))
+	// If PortForward is set to true, and the deployment_type is k8s, start
+	// forwarding traffic to/from kafka to localhost.
+	if cfg.PortForward && dt == `"k8s"` {
+		var err error
+		runOnce.Do(func() {
+			// NOTE(marclop) These assume terraform uses the strimzi operator.
+			if cfg.PortForwardResource == "" {
+				cfg.PortForwardResource = fmt.Sprintf("service/%s-kafka-bootstrap", cfg.Name)
+			}
+			if cfg.PortForwardMapping == "" {
+				cfg.PortForwardMapping = "9093:9093"
+			}
+			err = kubectlForwardKafka(cfg.Namespace,
+				cfg.Name,
+				cfg.PortForwardResource,
+				cfg.PortForwardMapping,
+			)
+		})
+		if err != nil {
+			return err
+		}
+	}
+	logger().Info("Kafka infastructure fully provisioned!")
+	return nil
+}
+
+func kubectlForwardKafka(ns, name, service, mapping string) error {
+	cmd := exec.Command("kubectl", "port-forward", "-n", ns, service, mapping)
+	logger().Infof("Running %s...", strings.Join(cmd.Args, " "))
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	// Patch zone for Kafka client calls to be correctly resolved.
+	srv, err := mockdns.NewServerWithLogger(map[string]mockdns.Zone{
+		// NOTE(marclop) Assumes terraform uses the strimzi operator.
+		fmt.Sprintf("%s-kafka-0.%s-kafka-brokers.%s.svc.", name, name, ns): {
+			A: []string{"127.0.0.1"},
+		},
+	}, log.New(io.Discard, "", 0), false)
+	if err != nil {
+		return err
+	}
+	if mockResolver == nil {
+		resolver := new(net.Resolver)
+		srv.PatchNet(resolver)
+		mockResolver = resolver
+	}
+	go cmd.Wait() // Run the forwarder
+	RegisterDestroy(ns+service+mapping, func() {
+		logger().Infof("Stopping %s...", strings.Join(cmd.Args, " "))
+		srv.Close()
+		cmd.Process.Kill()
+		cmd.Process.Wait()
+	})
+	return nil
+}
+
+func newKafkaTLSDialer() *tls.Dialer {
+	return &tls.Dialer{
+		NetDialer: &net.Dialer{Resolver: mockResolver},
+		Config:    &tls.Config{InsecureSkipVerify: true},
+	}
+}
+
+// KafkaBrokers returns the Kafka brokers to use for tests.
+func KafkaBrokers() []string {
+	brokersMu.RLock()
+	defer brokersMu.RUnlock()
+	return append(kafkaBrokers)
```

Comment:
What is the purpose of using `append` here?

Reply:
The idea is to return a copy of the slice, so if clients modify the slice, it doesn't have side effects.

Reply:
`append` does not create a new array, it modifies the current one. So using it with a single argument does **not** copy the slice.
https://go.dev/blog/slices-intro

Reply:
yep, fixed [2f7717b](https://github.com/elastic/apm-queue/pull/99/commits/2f7717b5dd32a15c0a68134d04e442aaaf278c9d)